cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -82,23 +88,81 @@ pub struct CheckedBuildMetadata {
     _private: (),
 }
 
+#[allow(async_fn_in_trait)]
+pub trait BinaryCache {
+    async fn upload(&self, path: &str) -> Result<(), PublishError>;
+    fn cache_uri(&self) -> &Url;
+}
+
+pub struct NixCopyCache {
+    pub uri: Url,
+    pub key_file: String,
+}
+
+impl BinaryCache for NixCopyCache {
+    async fn upload(&self, path: &str) -> Result<(), PublishError> {
+        debug!("Uploading {} to cache...", path);
+        let uri_with_key = format!("{}?secret-key={}", self.uri.as_str(), self.key_file);
+        let mut copy_command = Command::new(&*NIX_BIN);
+        copy_command
+            .arg("copy")
+            .arg("--to")
+            .arg(uri_with_key)
+            .arg(path);
+        let output = copy_command
+            .output()
+            .map_err(|e| PublishError::CacheUploadError(e.to_string()))?;
+        if output.status.success() {
+            Ok(())
+        } else {
+            let stderr = String::from_utf8_lossy(&output.stderr);
+            Err(PublishError::CacheUploadError(stderr.to_string()))
+        }
+    }
+
+    fn cache_uri(&self) -> &Url {
+        &self.uri
+    }
+}
+
+pub struct MockCache {
+    pub uri: Url,
+    pub error_msg: Option<String>,
+}
+
+impl BinaryCache for MockCache {
+    async fn upload(&self, _path: &str) -> Result<(), PublishError> {
+        if let Some(msg) = &self.error_msg {
+            Err(PublishError::CacheUploadError(msg.to_string()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn cache_uri(&self) -> &Url {
+        &self.uri
+    }
+}
+
 /// The `PublishProvider` is a concrete implementation of the `Publish` trait.
 /// It is responsible for the actual implementation of the `Publish` trait,
 /// i.e. the actual publishing of a package to a catalog.
 ///
 /// The `PublishProvider` is a generic struct, parameterized by a `Builder` type,
 /// to build packages before publishing.
-pub struct PublishProvider<Builder> {
+pub struct PublishProvider<Builder, Cache> {
     pub env_metadata: CheckedEnvironmentMetadata,
     pub build_metadata: CheckedBuildMetadata,
+    pub cache: Option<Cache>,
 
     pub _builder: Option<Builder>,
 }
```

Comment:
i'm not yet sure about generics vs, passing builder and cache to the `publish` method as `&impl Builder` and `&impl cache` (like we already do with `Client`).
I think i prefer the latter because you can reuse instances of the interfaces without splitting references between multiple subsystems.

Not something we need to get slowed down by now necessarily tho.